### ISSUE TITLE
Feature request: parameter to limit parallelism on aggregate steps

### ISSUE THREAD

**Issue by itsdalmo** [concourse/concourse#2628]
## What challenge are you facing?

I have a pipeline which runs tests on pull requests for a monolithic repository containing terraform. Initially I wanted to use `aggregate:` so that the execution time stayed the same no matter how many directories were being tested. But as the number of directories increased, we encountered issues with the number of sockets that were opened on a given worker (and failing to open new ones).

Part of the problem might be that we are using a monolithic repository which means that a single volume containing the source code is needed by all the tasks in the `aggregate:` step, which means they tend to all end up on the same worker (Using `random` container placement is not really a good solution for us, since most other pipelines don't have this problem and would all be penalised from not having reliable caches).

Currently we have to manually batch the tasks together (while trying to balance the number of tasks running in parallel, with the execution time of each task :joy:) to avoid the issue...

## What would make this better?

I'd like to be able to specify a large number of tasks in an `aggregate:` step, and be able to limit how many of them should be running at any given time until all of them complete (in other words, I want to be able to limit the maximum parallelism of an aggregate step).

For example:
```yml
- aggregate:
  max_parallelism: 5
  - task: A
  ...
  - task: Z
```

Would start 5 random tasks to begin with, and then keep exactly 5 tasks running until they all had completed. `max_parallelism` would be equivalent to `max_in_flight` for jobs (and could probably have the same name?).

## Are you interested in implementing this yourself?

Would love to contribute, but I would need some hand-holding to even figure out where to start looking to implement these changes 😅

**Comment by vito** [comment]
Cool idea!

This would be a change to the [`aggregate` step](https://github.com/concourse/concourse/blob/master/atc/exec/aggregate.go).

Alternatively - and this is my own itch to scratch - I'd like to start phasing out `aggregate` because its name is kind of stupid and confusing when used for tasks. It's a holdover from when being placed in an `aggregate` actually had an effect on the file layout that led to a `task`. Which is, like, pre-1.0.

A `parallel` step which supports this could be the start of transitioning away from it.

**Comment by itsdalmo** [comment]
Thanks!

I've been playing around with this right now, and I have a `parallel` step which works. However, the initial suggestion above is not valid YAML 😅  What I have now looks like this:

```yaml
jobs:
  - name: test-parallel
    plan:
    - get: every-minute
      trigger: true
    - parallel:
      - task: echo
        config:
         ...
      - task: echo
        config:
          ...
      max_in_parallel: 1
```

I.e. `max_in_parallel` has to be specified at the end of the block in order for it to be valid YAML, which looks a bit funky since preferably the constraints would be specified up top. Some options that might be better (collapsed the less relevant ones):

<details><summary>Specify a `name:` for the parallel step</summary>
<p>

```yaml
jobs:
  - name: test-parallel
    plan:
    - get: every-minute
      trigger: true
    - name: some-parallel-work
      max_in_parallel: 1
      parallel:
      - task: echo
        config:
         ...
      - task: echo
        config:
          ...
```

Does not make a lot of sense since we won't use the name anywhere.

</p>
</details>

<details><summary>Add a Parallel struct to `atc.PlanConfig`</summary>
<p>

If we added a struct like the below the `atc.PlanConfig`:

```go
type ParallelConfig struct {
  Steps *PlanSequence
  Limit int
}
```

We could have a YAML specification like this:

```yaml
jobs:
  - name: test-parallel
    plan:
    - get: every-minute
      trigger: true
    - parallel:
        limit: 1 # or same name "max_in_parallel"
        steps: # or another name like "sequence"
        - task: echo
          config:
           ...
        - task: echo
          config:
            ...
```

This to me would be the most correct implementation, but I don't like the indentation of `limit` and `steps` - and it would be the first `step` to require a special indentation? 

</p>
</details>

Reusing either `do` or `aggregate` would also work, since they can be made to be the default behaviour for `parallel`. So we could do something like the following:

```yaml
jobs:
  - name: example
    plan:
    - parallel: 3 # Default to 1 when omitted, i.e. regular "do"
      do: 
      - task: echo
        config:
         ...

  - name: example-2
    plan:
    - parallel: 2 # Default to length of plan, i.e. regular "aggregate"
      aggregate: 
      - task: echo
        config:
         ...
```

Alternately we just add a new key `sequence` (or `steps`):

```yaml
jobs:
  - name: example
    plan:
    - parallel: 3 # Runs 3 at a time in parallel
      sequence: 
      - task: echo
        config:
         ...
    - sequence: # Runs 1 at a time (default behavior)
      - task: echo
        config:
         ...
    - parallel: # aggregate. If we can tell the difference between this and the above. Perhaps using a pointer to an int works? So it's nil if it was not specified, and 0 if the key is present with no value?
      sequence:
      - task: echo
        config:
         ...
```

#### How would you like the PR?

I've grokked something which works (top YAML block) by essentially copy-pasting `aggregate` code and making small adjustments to make it work like what is proposed for the `parallel` step. However, this makes for a very large `diff` since tests have to be copied also, even though they are more or less identical.

So I'm wondering if perhaps you'd instead want me to adjust [exec.AggregateStep](https://github.com/concourse/concourse/blob/master/atc/exec/aggregate.go) so that it supports limiting the parallelism, setting the limit to equal the length of the `[]Step`'s by default (so `aggregate` works as normal),  and basing `parallel` off the modified `exec.AggregateStep` instead?

tl;dr How do you want the `parallel` step to look in YAML, and how do you want the PR @vito? 😂

**Comment by vito** [comment]
@itsdalmo Sorry I never replied, thanks for the PR! Got caught up in getting 5.0.1 out. I'll take a look at the PR now.
